keps/sig-api-machinery/5647-stale-controller-handling/README.md

```diff
@@ -0,0 +1,811 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+We propose a way to mitigate staleness within Kubernetes controllers,
+particularly the ones located in `kube-controller-manager` (KCM). When operating
+at a large scale, controllers can fall behind the state of the apiserver,
+leading to decisions based on outdated information. This proposal introduces a
+generalized mechanism for controllers to detect how stale their local cache is
+and prevent reconciliation until the cache catches up to provide consistency.
+
+## Motivation
+
+Every controller operates on a local cache that is populated by watching for
+changes from the apiserver. By its nature, this watch stream is eventually
+consistent and provides no guarantee of how far behind the "live" state of the
+apiserver it is. A change event might arrive within milliseconds, or under other
+circumstances, could be delayed by seconds or even minutes. This means every
+controller, at all times, is operating on a potentially outdated view of the
+cluster's state. 
+
+The issue this KEP addresses is that operators currently have no visibility into
+this lag. It's impossible to distinguish between normal delays and a controller
+that has fallen dangerously out of sync. A controller may continue trying to
+enact its desired state based on an outdated view of the world, causing spurious
+reconciles that do not reflect the desired intentions of the user. 
+
+This proposal aims to provide solutions to mitigate issues that may arise out of
+the controller falling behind. With the ability to [compare resource
+version](https://github.com/kubernetes/enhancements/pull/5505) we now have the
+ability to have the idea of the "age" of a resource and run operations with that
+in mind.
+
+### Goals
+
+The goals of this KEP is to document the stale controller issue and propose
+solutions to help detect and mitigate it. We will propose the ideas of metrics
+and enforcing stricter consistency semantics at certain points in the reconcile
+cycle of a controller.
+
+### Non-Goals
+
+This is not intended to enforce consistency guarantees on all controllers. The
+changes described here will be "opt-in" for every controller and will require
+some degree of changes to the controller logic. It is not the goal to require
+controllers to all view a consistent state of the world per reconcile.
+
+## Proposal
+
+We propose the ability for controllers to be able to view whether their own
+writes have made it into the apiserver. This will allow for the controller to
+pause their reconcile loop until it knows whether or not the previous write has
+been propogated to its cache. We will do this by adding plumbing to the
+underlying cached informer and adding the ability for controllers to know the
+resource version of objects they subscribe to. 
```

Comment:
I think that in majority of cases, I don't necessarily want to pause the whole controller (that would effectively make it single-threaded), but rather I want to pause it "per-object" (i.e. I don't want to reconcile the object X again, before I observed the changes associated with that object in my cache. For example: if by processing RS X I created pod1 and pod2, I don't want to reconcile RS again until pod1 and pod2 are already observed in my cache).

At least for me the text is slightly misleading and I would suggest clarifying this a bit.